kernel/src/log_segment.rs

```diff
@@ -104,13 +113,240 @@ impl LogSegment {
+        let (mut commit_files, mut checkpoint_files) = match (checkpoint, end_version) {
+            (Some(cp), None) => Self::list_log_files_with_checkpoint(&cp, fs_client, &log_url)?,
+            (Some(cp), Some(version)) if cp.version >= version => {
+                Self::list_log_files_with_checkpoint(&cp, fs_client, &log_url)?
+            }
+            _ => {
+                let (commit_files, checkpoint_files, _) =
+                    Self::list_log_files_from_version(fs_client, &log_url, None)?;
+
+                (commit_files, checkpoint_files)
+            }
+        };
+
+        if no_checkpoint_files {
+            checkpoint_files.clear();
+        }
+
+        // Commit file versions must satisfy the following:
+        // - Be greater than the start version
+        // - Be greater than the most recent checkpoint version if it exists
+        // - Be less than or equal to the end version.
+        if let Some(start_version) = start_version {
+            commit_files.retain(|log_path| start_version <= log_path.version);
+        }
+        if let Some(checkpoint_file) = checkpoint_files.first() {
+            commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
+        }
+        if let Some(end_version) = end_version {
+            commit_files.retain(|log_path| log_path.version <= end_version);
+        }
+
+        // get the effective version from chosen files
+        let version_eff = commit_files
+            .last()
+            .or(checkpoint_files.first())
+            .ok_or(Error::MissingVersion)? // TODO: A more descriptive error
+            .version;
+        if let Some(end_version) = end_version {
+            require!(
+                version_eff == end_version,
+                Error::MissingVersion // TODO more descriptive error
+            );
+        }
+
+        // We assume listing returned ordered. If `in_order_commit_files` is false, we want reverse order.
+        if !in_order_commit_files {
+            commit_files.reverse();
+        }
+
+        Ok(LogSegment {
+            end_version: version_eff,
+            log_root: log_url,
+            commit_files: commit_files
+                .into_iter()
+                .map(|log_path| log_path.location)
+                .collect(),
+            checkpoint_files: checkpoint_files
+                .into_iter()
+                .map(|log_path| log_path.location)
+                .collect(),
+        })
+    }
+    pub(crate) fn list_log_files_from_version(
+        fs_client: &dyn FileSystemClient,
+        log_root: &Url,
+        version: Option<Version>,
+    ) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>, i64)> {
+        let begin_version = version.unwrap_or(0);
+        let version_prefix = format!("{:020}", begin_version);
+        let start_from = log_root.join(&version_prefix)?;
+
+        let mut max_checkpoint_version = version.map_or(-1, |x| x as i64);
```

Comment:
Is the `max_checkpoint_version = -1` sensible? This is what we used to do in the checkpoint case, but perhaps we want to rethink this?

Reply:
why not just `Option<Version>`?

Reply:
should we rename to `checkpoint_version`? and probably have docs that says that we are really only tracking one checkpoint (the highest) and N commits